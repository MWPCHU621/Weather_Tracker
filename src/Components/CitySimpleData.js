import { useDispatch } from 'react-redux';
import { addWeatherInfo } from '../Reducers/WeatherInfoSlice';
import { getCityInfo, getCityInfoFiveDay } from '../Helper/ApiCalls';
import { getWeatherDayInfo, weatherToIcon } from '../Helper/HelperFunctions';
import { removeCity, refreshCityData} from '../Reducers/CityListSlice';
import { Button } from '@material-ui/core';
import RefreshIcon from '@material-ui/icons/Refresh';
import ClearIcon from '@material-ui/icons/Clear';
import '../style/citySimpleData.css';

export function CitySimpleData(props) {

    const dispatch = useDispatch();
    const city = props;

    return(
        <div className="simple_city_data">
            <Button 
                className="city_text" 
                onClick={() => getCityDetailedInfo(city.name)}
                style={{justifyContent:"flex-start", backgroundColor:"transparent", fontSize:"large"}}
                disableRipple
            > 
                {city.name} - {city.temperature} {weatherToIcon(city.weather)} 
            </Button>
            <Button className="city_refresh" size="small" onClick={() => refreshCity(city.name)}><RefreshIcon /></Button>
            <Button className="city_remove" onClick={() => remove(city)}><ClearIcon /></Button>
        </div>
    )

    //gets 5 days worth of weather info for given city
    function getCityDetailedInfo(city) {
        getCityInfoFiveDay(city)
        .then(data => {
            let newData = {
                name: data.city.name,
                currentTemp: (Math.round(data.list[0].temp.day) + "C"),
                currentWeather: data.list[0].weather[0].main,
                description: data.list[0].weather[0].description,
                wind: ( Math.round(data.list[0].speed) + "ms " + data.list[0].deg + " deg"),
                pressure: data.list[0].pressure,
                dailyWeatherData: [],
            }

            //loop to create and push modified daily info to newData to be added to store.
            for(let i=0; i<5; i++) {
                newData.dailyWeatherData.push(getWeatherDayInfo(data.list[i]));
            }
            dispatch(addWeatherInfo(newData));
        })
    }

    //refreshes the simple weather info for a given city
    function refreshCity(cityName) {
        getCityInfo(cityName)
        .then(data => {
            // creating new object to be stored in city List for display in left container.
            let cityData = {
                name: data.name,
                weather: data.weather[0].description,
                temperature: (Math.round(data.main.temp) + "C"),
            }
            dispatch(refreshCityData(cityData));
        })
    }

    //removes a single given city from the list 
    function remove(city) {
        dispatch(removeCity(city));
    }
}