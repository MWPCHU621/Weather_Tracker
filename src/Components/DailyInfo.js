import { weatherToIcon } from '../Helper/HelperFunctions';

export function DailyInfo(props) {

    const dailyInfo = props.dailyInfo;
    
    return(
        <div className="daily_weather_info">
            <div>{dailyInfo.date}</div>
            <div>{dailyInfo.dayOfWeek}</div>
            <div> { weatherToIcon(dailyInfo.weather) } </div>
            <div>{dailyInfo.temp}</div>
        </div>
    )
}