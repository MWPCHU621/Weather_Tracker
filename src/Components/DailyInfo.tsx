import { weatherToIcon } from '../Helper/HelperFunctions';
import '../style/dailyInfo.css';

export function DailyInfo(props: any): JSX.Element {

    const dailyInfo = props.dailyInfo;
    
    return(
        <div className="daily_weather_info">
            <div>{dailyInfo.date}</div>
            <div>{dailyInfo.dayOfWeek}</div>
            <div> { weatherToIcon(dailyInfo.weather) } </div>
            <div>{dailyInfo.weather}</div>
            <div>{dailyInfo.temp}</div>
        </div>
    )
}