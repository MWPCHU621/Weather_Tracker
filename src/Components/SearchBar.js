import { useDispatch } from 'react-redux';
import { useState } from 'react';
import {addCity} from '../Reducers/SearchBarSlice';
import { getCityInfo } from '../Helper/ApiCalls';

import SearchBar from 'material-ui-search-bar';


export function CitySearchBar() {
    const dispatch = useDispatch();

    const [value, setValue] = useState("");

    return (
        <div>
            <div className="searchBar_container">
                <SearchBar
                    value={value}
                    onChange={(newValue) => setValue(newValue)}
                    onRequestSearch={() => { searchCity(value); setValue(""); }} 
                />
                <hr className="horizontal_line"></hr>
            </div>
        </div>
    );

    //will remove once above form is swapped for form from material UI
    function searchCity (city) {
        getCityInfo(city).then((data) => {
            let cityData = {
                name: data.name,
                weather: data.weather[0].description,
                temperature: (Math.round(data.main.temp) + "C"),
            }
    
            dispatch(addCity(cityData));
        })
        .catch(error => {
            alert("Invalid City. Please enter a valid city");
        })
    }

}