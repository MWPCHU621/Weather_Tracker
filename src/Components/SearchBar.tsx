import { useDispatch } from 'react-redux';
import { useState } from 'react';
import { convertCityToCoord, convertZipcodeToCoord, getCityInfoSevenDay } from '../Helper/ApiCalls';
import AddIcon from '@material-ui/icons/Add';
import { TextField } from '@material-ui/core';
import '../style/searchBar.css';
import { addWeatherInfo } from '../Reducers/WeatherInfoSlice';
import { getWeatherDayInfo } from '../Helper/HelperFunctions';

export function CitySearchBar(): JSX.Element {

    const dispatch = useDispatch();
    const [cityName, setCityName] = useState("");
    const [error, setError] = useState(false);
    const [zipCountry, setZipcode] = useState("");
    
    return (
        <div className="searchBar_container">
            <form onSubmit={handleCitySubmit} className="add_city_form">
                <TextField className="search_city_textarea textarea" placeholder="City Name" label="Search with city name"value={cityName} onChange={handleCityChange}/>
                <button type="submit" className="add_city_btn"><AddIcon /></button>
            </form>

            <form onSubmit={handleZipcodeSubmit} className="add_city_form">
                <TextField 
                    className="search_zipcode_textarea textarea" 
                    placeholder="a2c,ca" 
                    label="Search with zipcode, country code"
                    value={zipCountry} 
                    onChange={handleZipcodeChange} 
                />
                <button type="submit" className="add_city_btn"> <AddIcon /> </button>
            </form>

            <p className={error ? 'show_error' : 'hide_error'}>Invalid city or zipcode country code combination. Please enter a valid input</p>
        </div>
    );

    function handleCityChange(e: any) {
        setCityName(e.target.value);
    }

    function handleZipcodeChange(e: any) {
        setZipcode(e.target.value);
    }

    function handleCitySubmit(e: any) {
        e.preventDefault();
        searchCity(cityName);
    }

    function handleZipcodeSubmit(e: any) {
        e.preventDefault();
        searchZipCode(zipCountry);
    }
    
    function searchCity (city: string) {
        convertCityToCoord(city).then((data) => {
            const lat = data[0].lat.toFixed(2);
            const lon = data[0].lon.toFixed(2);
            const coord = {
                lat: lat,
                lon: lon
            };

            return coord;
        })
        .then(coord => {
            getCityInfoSevenDay(coord.lat, coord.lon).then((data) => {
                let weatherInfo: any = {
                    description: data.current.weather[0].description,
                    currentTemp: data.current.feels_like + "C",
                    minTemp: data.daily[0].temp.min + "C",
                    maxTemp: data.daily[0].temp.max + "C",
                    windSpd: data.current.wind_speed,
                    precipitation: (data.daily[0].rain == undefined) ? 0 : data.daily[0].rain,
                    humidity:data.current.humidity,
                    dailyInfo: [],
                }
                
                for(let i=1; i<8; i++) {
                    weatherInfo.dailyInfo.push(getWeatherDayInfo(data.daily[i]));    
                }
                
                setError(false);
                dispatch(addWeatherInfo(weatherInfo));
            })
            .catch(error => setError(true));
            
        })
        .catch(error => setError(true));

    }

    function searchZipCode(zipCountry: string) {
        const strSplit = zipCountry.split(",");
        const zipcode = strSplit[0];
        const countryCode = strSplit[1];

        convertZipcodeToCoord(zipcode, countryCode).then(data => {
            const lat = data.lat.toFixed(2);
            const lon = data.lon.toFixed(2);

            const coord={
                lat:lat,
                lon:lon
            };

            return coord;
        })
        .then(coord => {
            getCityInfoSevenDay(coord.lat, coord.lon).then((data) => {
                let weatherInfo: any = {
                    description: data.current.weather[0].description,
                    currentTemp: data.current.feels_like,
                    minTemp: data.daily[0].temp.min,
                    maxTemp: data.daily[0].temp.max,
                    windSpd: data.current.wind_speed,
                    precipitation:data.daily[0].rain,
                    humidity:data.current.humidity,
                    dailyInfo: [],
                }

                for(let i=1; i<8; i++) {
                    weatherInfo.dailyInfo.push(getWeatherDayInfo(data.daily[i]));    
                }
                
                setError(false);
                dispatch(addWeatherInfo(weatherInfo));
               
            })
            .catch(error => setError(true));
        })
        .catch(error => setError(true));
    }
}