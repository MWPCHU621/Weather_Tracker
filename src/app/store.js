import { configureStore } from '@reduxjs/toolkit';
import cityListReducer from '../Reducers/SearchBarSlice';

export const store = configureStore({
  reducer: {
    cities: cityListReducer,
    weatherInfo: {},
  },
});
