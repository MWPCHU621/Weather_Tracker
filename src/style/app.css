.App {
    width: 100%;
    height:100%;
    display: flex;
    flex-direction: row;
    justify-content:space-around;
    margin-top: 20px;
}

.left_container {
    margin-right: 10px;
    flex:35%;
    height:80%;
    border:2px solid black;
    display:flex;
    flex-direction: column;
}

.right_container {
    flex:65%;
    height:80%;
    border:2px solid black;
}