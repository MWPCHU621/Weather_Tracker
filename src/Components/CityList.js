import { useSelector, useDispatch } from 'react-redux';

import {getCityList} from '../Reducers/SearchBarSlice';
import { CitySimpleData } from './CitySimpleData';

import { Button } from '@material-ui/core';


export function CityList() {
    // const cityList = useSelector(getCityList);
    const dispatch = useDispatch();

    const cityList = useSelector(getCityList);

    return (
        <div>
            <div className="city_list_container">
                {
                    console.log("cityList", cityList),
                    cityList.map((city, index) => {
                        return (
                            <CitySimpleData key={index} {...city}/>
                        )
                    })
                }
            </div>
        </div>
    );

}