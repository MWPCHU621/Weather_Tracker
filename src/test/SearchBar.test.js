// Link.react.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { CitySearchBar } from '../Components/SearchBar';

test('Link changes the class when hovered', () => {
    const component = renderer.create(
        <Provider store={store}>
            <CitySearchBar />,
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});