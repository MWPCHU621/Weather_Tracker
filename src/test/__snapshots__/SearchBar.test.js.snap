// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
Array [
  <div
    className="searchBar_container"
  >
    <div
      className="MuiPaper-root ForwardRef-root-1 MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "boxShadow": "none",
        }
      }
    >
      <div
        className="ForwardRef-searchContainer-7"
      >
        <div
          className="MuiInputBase-root MuiInput-root ForwardRef-input-6 MuiInputBase-fullWidth MuiInput-fullWidth"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="Type City Name"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiIconButton-root ForwardRef-iconButton-2 ForwardRef-searchIconButton-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root ForwardRef-icon-5"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root ForwardRef-iconButton-2 ForwardRef-iconButtonHidden-3"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root ForwardRef-icon-5"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
      </button>
    </div>
    <p
      className="hide_error"
    >
      Invalid city. Please enter a valid city
    </p>
  </div>,
  ",",
]
`;
